### PR TITLE
Bug 1143928 - Refactor Deck and Panel classes to reduce code duplicat…

### PR DESCRIPTION
**Pull request by utvar**
…ion. r=whimboo

### CODE REVIEW

**Comment by utvar** [comment]
Work in Progress.

File: firefox_puppeteer/ui/deck.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from marionette_driver import By
+
+from firefox_puppeteer.ui_base_lib import UIBaseLib
+
+
+class BaseDeck(UIBaseLib):
```
Comment (whimboo) [inline firefox_puppeteer/ui/deck.py]: I would keep `Deck` as the name of the base class and change the name of the specific subclass.

File: firefox_puppeteer/ui/deck.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                   'checkingForUpdates': CheckingForUpdatesPanel,
+                   'downloadAndInstall': DownloadAndInstallPanel,
+                   'downloadFailed': DownloadFailedPanel,
+                   'downloading': DownloadingPanel,
+                   'feedPanel': FeedPanel,
+                   'generalPanel': GeneralPanel,
+                   'mediaPanel': MediaPanel,
+                   'noUpdatesFound': NoUpdatesFoundPanel,
+                   'permPanel': PermissionsPanel,
+                   'securityPanel': SecurityPanel
```
Comment (whimboo) [inline firefox_puppeteer/ui/deck.py]: All those panels are specific classes for the update deck in the about window. So they have to stay as part of the subclass in the other file. As of now the class references cannot be found.

File: firefox_puppeteer/ui/deck.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                   }
+
+        panel = self.element.find_element(By.ID, panel_id)
+        return mapping.get(panel_id, BasePanel)(lambda: self.marionette, self.window, panel)
+
+
+class BasePanel(UIBaseLib):
```
Comment (whimboo) [inline firefox_puppeteer/ui/deck.py]: Same here. Keep the name as we had before for the base class.

File: firefox_puppeteer/ui/deck.py (hunk 2 above)
Comment (utvar) [inline firefox_puppeteer/ui/deck.py]: @whimboo, I have still not found a way to extract the _create_panel_for_id method without also pulling out all the panel classes. It seems the only things remaining in Deck that are common to both about_window and pageinfo, that can be extracted without also pulling out anything specific to one of these two, are three properties: panels, selected_index, and selected_panel. (These three are also in update_wizard, inside the Wizard class.) Is this what you would like refactored out?

Comment (whimboo) [inline firefox_puppeteer/ui/deck.py]: `_create_panel_for_id()` can be made an abstract method so subclasses have to implement their own mapping logic.

Even better would be to create the mapping object at class level, so the `_create_panel_for_id()` method in the base class could access it via e.g. `self.panel_map`. With that each subclass would only have to create this object and everything else happens magically.

**Comment by utvar** [comment]
@whimboo, not sure why the one Linux build failed -- I looked at the errors and am fairly sure it is not my code. The tests pass locally using firefox-ui-functional. (Using firefox-ui-tests returns ImportError: 'module' object has no attribute 'cli', for the mozilla-central branch as well.)

I'm not sure what will happen to this PR now that the code has been moved, but I at least wanted to return my code to a working state, so Deck is back where it was and only Panel has been refactored out.

**Comment by whimboo** [comment]
This repository stays as the canonical repository for now. At least until we can do per checkin tests in hg.mozilla.org/mozilla-central.

The failing test here is a known intermittent failure I was never able to reproduce myself. So don't worry about it. The rest of this PR would be ready for another review now?

**Comment by utvar** [comment]
@whimboo, Travis results are mixed, but as far as I can tell, it is not related to my code.

**Comment by whimboo** [comment]
This looks pretty fine for the base Panel class. Do you want to fold in the refactoring of the base Deck class in the same pull or do that on another one? I'm open for both.

**Comment by utvar** [comment]
@whimboo, I'd prefer to do a separate PR for the Deck class, because I'm not sure how long it will take me to finish that one.

**Comment by whimboo** [comment]
Alright, I see. I will have a closer look at this PR now.

**Comment by whimboo** [comment]
The code changes look fine and are at the right places. But I noticed that we do not have a rst file for the documentation yet. Please add one for the new ui module under `firefox_puppeteer/docs/ui/`.

**Comment by utvar** [comment]
@whimboo, here is the addition you requested.

**Comment by whimboo** [comment]
I think that looks fine now. I will get it merged. Thanks @utvar!

**Comment by whimboo** [comment]
This PR got merged as https://github.com/mozilla/firefox-ui-tests/commit/023fb06a01b86d56363a0dc46c2317a79aa09570.

**Pull request closed without merging**